Join all hammer output parsers in one module

elyezer opened the pull request:
Move hammer output parsers from the common.helpers module to cli.hammer
module and add a new parse for help output. Also add and update tests
for the new changes.

The parser for help output will help implement #1862.

elyezer commented:
Update: rebased

Ichimonji10 commented on robottelo/cli/hammer.py:
```diff
@@ -0,0 +1,163 @@
+
+        if state == subcommands_section_state:
+            match = subcommand_regex.search(line)
+            if match is not None:
```
Can we make the following change to the logic of this function?

```python
if match is None:
    continue
if match.group('name') is None:
   …
else:
    …
```

I find logic easier to follow when "error" cases are checked for and dealt with than when "success" cases are checked for and dealt with. It tends to produce less indented code, if nothing else.

Ichimonji10 replied:
The same remark applies to line 70.

elyezer replied:
:+1:

Ichimonji10 commented on robottelo/cli/hammer.py:
```diff
@@ -0,0 +1,163 @@
+
+
+def parse_help(output):
+    """Parse the help output from a hammer command and return a dictinary
```
`s/dictinary/dictionary/`

elyezer replied:
thanks

Ichimonji10 commented on robottelo/cli/hammer.py:
```diff
@@ -0,0 +1,163 @@
+def parse_info(output):
+    """Parse the info output and returns a dict mapping the values."""
+    # info dictionary
+    r = {}
```
Can we use a more informative variable name than `r`?

elyezer replied:
Just moved this function to later refactor it, but will change this var name for now and will refactor other things later

Ichimonji10 replied:
Sounds good.

Ichimonji10 commented:
ACK pending comments.

Can you provide a brief example of how this all works together? I see unit tests, which is nice, but a quick summary of the intended usage of this code would be fantastic.

elyezer commented:
The CSV and info parser was already created and they are used to parse the output provided by hammer and convert it in a format that Robottelo tools can work with. The help parser was added in this PR in order to allow grab the information about the available subcommands and options for a given hammer command.

Sample usage can be like below. The strings was split on `\n` because this is the way that the ssh module leave the output.

```python
>>> from robottelo.cli import hammer
>>> csv_output = '''Message,Id,Name
... Organization created,329,RekMxN
...
... '''
>>> csv_output = csv_output.split('\n')
>>> hammer.parse_csv(csv_output)
[{'message': 'Organization created', 'id': '329', 'name': 'RekMxN'}]
```

```python
>>> info_output = '''ID:                          342
... Name:                        56GBrH
... Label:                       56GBrH
... Description:
... Organization:                RekMxN
... Library:                     false
... Prior Lifecycle Environment: Library
...
... '''
>>> info_output = info_output.split('\n')
>>> hammer.parse_info(info_output)
{'description': {}, 'library': 'false', 'label': '56GBrH', 'prior-lifecycle-environment': 'Library', 'organization': 'RekMxN', 'id': '342', 'name': '56GBrH'}
```

```python
>>> help_output = '''Usage:
...     hammer activation-key add-subscription [OPTIONS]
...
... Options:
...  --id ID                                 ID of the activation key
...  --name NAME                             Activation key name to search by
...  --organization ORGANIZATION_NAME        Organization name to search by
...  --organization-id ORGANIZATION_ID       organization ID
...  --organization-label ORGANIZATION_LABEL Organization label to search by
...  --quantity QUANTITY                     Quantity of this subscription to add
...  --subscription-id SUBSCRIPTION_ID       ID of subscription
...  -h, --help                              print help
...
... '''
>>> help_output = help_output.split('\n')
>>> hammer.parse_help(help_output)
{'subcommands': [], 'options': [{u'shortname': None, u'name': 'id', u'value': 'ID', u'help': 'ID of the activation key'}, {u'shortname': None, u'name': 'name', u'value': 'NAME', u'help': 'Activation key name to search by'}, {u'shortname': None, u'name': 'organization', u'value': 'ORGANIZATION_NAME', u'help': 'Organization name to search by'}, {u'shortname': None, u'name': 'organization-id', u'value': 'ORGANIZATION_ID', u'help': 'organization ID'}, {u'shortname': None, u'name': 'organization-label', u'value': 'ORGANIZATION_LABEL', u'help': 'Organization label to search by'}, {u'shortname': None, u'name': 'quantity', u'value': 'QUANTITY', u'help': 'Quantity of this subscription to add'}, {u'shortname': None, u'name': 'subscription-id', u'value': 'SUBSCRIPTION_ID', u'help': 'ID of subscription'}, {u'shortname': 'h', u'name': 'help', u'value': None, u'help': 'print help'}]}
```

Ichimonji10 commented:
That is *fantastic*. Thank you.

elyezer commented:
@Ichimonji10 glad you like it. Thanks for the review.

omaciel commented:
ACK
